### ISSUE TITLE
OOM in Quarkus 3.0.0.Beta1 caused by okio via OpenTelemetry

### ISSUE THREAD

**Issue by snazy** [quarkusio/quarkus#32238]
### Describe the bug

okio starts a thread via the class `okio.AsyncTimeout.Watchdog` once "it has to deal with a timeout". The thread's configured as a daemon thread and seems to have some shutdown logic, and also aborts when its interrupted.

The behavior isn't an issue in production usecases, but it's an issue when running tests.

`okio.AsyncTimeout.Watchdog` is loaded by a Quarkus class loader for every test, so it implicitly keeps a reference to its class loader and transitively to all the resources that one holds - just because the thread's still running.

Setting the following parameters disables timeouts and in turn does not start that watchdog thread and the OOM doesn't happen. It's maybe a legit workaround for tests, until the issue's fixed.
```
quarkus.otel.exporter.otlp.timeout=0
quarkus.otel.exporter.otlp.traces.timeout=0
```

This behavior seems to be introduced after Quarkus 3.0.0.Alpha5, but I'm not sure why, because okio seems to behave this way "forever".

**Comment by brunobat** [comment]
What are the memory definitions that you use @snazy ?

**Comment by snazy** [comment]
At max 1.5G heap IIRC, maybe less

**Comment by brunobat** [comment]
I cannot run the tests. Throws me a too many open files before I reach to any OOM.
I'm in a locked mac. No way to increase more of that. 
Is there a specific test where the OOM will happen more often?
(will find a different, slower machine)

**Comment by snazy** [comment]
There are not that many test classes, OOMs pretty much „at the end“.

**Comment by brunobat** [comment]
I asked @alesj to take a look.

**Comment by alesj** [comment]
I already got this on the `main` branch
* https://gist.github.com/alesj/6f360b7c6eac7e12820cf0dedcf5028b

**Comment by alesj** [comment]
Where as this never stopped: https://gist.github.com/alesj/60d7467f85644dd639869f3c23833818 ... 42min and it would go on ...

@snazy any easier way to reproduce this -- not to have the whole Nessie project ... ?

**Comment by snazy** [comment]
Probably - you could create a test project with many test classes, and each produces an otel event.

**Comment by alesj** [comment]
@snazy what if you set this
* https://quarkus.io/guides/class-loading-reference#quarkus-class-loading-configuration-class-loading-config_quarkus.class-loading.parent-first-artifacts
for the `okio` library.
This way it will only load `okio` classes once for all tests.

**Comment by alesj** [comment]
@snazy any progress / luck with this `parent-first`?
Or easier reproducer ...

**Comment by snazy** [comment]
Sorry, no luck so far. It's pretty clear that it's caused by okio, already known and was hit before. I think, the current idea is to remove okio. It's "only" a test-issue (prod-like doesn't reload stuff).

**Comment by brunobat** [comment]
Yes and not just for this reason. We need to drop OkHttp because of support reasons.

**Comment by geoand** [comment]
Good point

**Comment by geoand** [comment]
So I assume this is one for @alesj due to the need for gRPC

**Comment by brunobat** [comment]
Yes, it would be nice

**Comment by geoand** [comment]
I might be misreading `OkHttpGrpcExporter`, but it seems to simply be making a HTTP call. Is that correct or am I missing something?

**Comment by brunobat** [comment]
The OkHttp lib is able to do both GRPC and HTTP, there is a config for that.

**Comment by geoand** [comment]
Yeah, I see that. My point is that in reality, `OkHttpGrpcExporter` doesn't really need to know much about gRPC.

And actually this is true, as the [description](https://github.com/open-telemetry/opentelemetry-java/pull/3684#issue-1011723505) of the original PR that introduced the class above says.

**Comment by brunobat** [comment]
Correct.
Also, beware of this: https://github.com/open-telemetry/opentelemetry-java/pull/5557/
We might not need to implement anything.

**Comment by geoand** [comment]
I saw that too, yeah. But from a quick read through it, doesn't seem to change any gRPC related stuff.

**Comment by geoand** [comment]
According to [this](https://github.com/open-telemetry/opentelemetry-java/issues/5351#issuecomment-1625277197), it's not possible to use the JDK HTTP Client because there is not trailing header support.

@cescoffier @alesj I assume we could leverage the Vert.x `GrpcClient` or even the `HttpClient`  here as all we need to do is shove some bytes over HTTP2 (see what they are [currently doing](https://github.com/open-telemetry/opentelemetry-java/blob/main/exporters/common/src/main/java/io/opentelemetry/exporter/internal/grpc/OkHttpGrpcExporter.java) with OkHttp). WDYT?

**Comment by brunobat** [comment]
That's actually my preference as well. 
Before this JDK HTTP Client implementation there was no API to plug a different client.
Everything was tight to OkHttp.
Now there is one.

**Comment by cescoffier** [comment]
What do you call trailing headers? trailer? gRPC is often using trailers, and we do support that.

**Comment by geoand** [comment]
@cescoffier yeah [trailer](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Trailer)

**Comment by cescoffier** [comment]
Yes, it should be fine.

**Comment by geoand** [comment]
I assume https://vertx.io/docs/vertx-grpc/java/#_message_level_api_2 is what we are after. I'll try and have a look next week

**Comment by geoand** [comment]
I have a prototype [here](https://github.com/quarkusio/quarkus/pull/34647)
